docs/proposals/remote-artifacts.md

```diff
@@ -0,0 +1,308 @@
+*   Create the mechanism to download and prepare remote artifacts for builds;
+
+## Non-Goals
+
+*   Automate the upload of local artifacts into the cluster;
+*   Manage remote artifacts;
+*   Amend container images created by the given `BuildStrategy`;
+*   Walking a remote directory tree, similarly to a `git checkout`;
+
+# Proposal
+
+The enhancement proposal is centered around the idea of declaring external dependencies, here
+called "artifacts", and being able to use them during the container image building process.
+
+## Build Artifacts
+
+The remote artifacts will be directly expressed in the body of a `Build` resource, as the
+following example:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: license-file
+spec:
+  sources:
+    - name: license-file
+      type: http
+      url: https://licenses.company.com/customer-id/license.tar
+      http:
+        path: $(workspace)/licenses/license.tar
+```
+
+We will add `sources` section, side by side with current `source`. The idea is to accommodate both
+constructions in `v1alpha1`, and save API breaking changes for upcoming `v1beta1`.
+
+The new `sources` will contain the following attributes:
+
+- `name`: source name (required);
+- `type`: input source type, `git` or `http` (required);
+- `sourceRef`: use a external resource, the name in this field is the resource name (optional);
+- `url`: the URL of the repository or remote artifact (optional);
+- `credentials`: the credentials to interact with the artifact (optional);
+- `http`: settings for `http`, namely `path` (optional);
+- `git`: settings for `git`, namely `revision` (optional);
+
+The resource `UID`/`GID` will be defined by the same user running the Tekton task, later on, we
+can extend the API to support arbitrary configuration.
+
+### Single vs. Multiple Git Repositories
+
+In `v1alpha1` it's expected to inform the Git repository on `.spec.source`, and external artifacts
+informed on `.spec.sources`, therefore, we can address adding this new feature and managing multiple
+Git repositories support in different tracks.
+
+For upcoming work, supporting multiple Git repositories will involve re-defining `/workspace/source`
+location and `$workspace` placeholder.
+
+### Standalone CRD
+
+Alternatively, we may define the artifacts as a standalone CRD, that is a `BuildSource`
+resource. The advantage of this design is being able to exchange, and reuse, artifacts on several
+builds. For instance, if two projects are sharing a common logo image, both `Builds` will refer to
+the same `BuildSource`.
+
+The following snippet shows how an Artifact (`BuildSource`) will be represented.
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: BuildSource
+metadata:
+  name: license-file
+spec:
+  sources:
+    - name: license.tar
+      type: http
+      url: https://licenses.company.com/customer-id/license.tar
+      http:
+        path: $(workspace)/licenses/license.tar
+```
```

Comment:
A question around the `BuildSource`. Within the `Build`'s `spec.sources`, it is used to from one source entry to reference an external definition. Based on that, may expectation would be that the external build source would only contain a single spec like this:

```yaml
---
apiVersion: build.dev/v1alpha1
kind: BuildSource
metadata:
  name: license-file
spec:
  type: http
  url: https://licenses.company.com/customer-id/license.tar
  http:
    path: $(workspace)/licenses/license.tar
```

Should we have multiple entries with their own name in the `BuildSource`, then we'll have to define how this is merged with the other entries in the `Build`'s sources.

Reply:
I think here we may benefit from a slice of entries. Given that we may be downloading a number of dependencies from a single external source, and we can bundle together in this slice fashion. It also keeps aligned with `Build` resource where it will be available as a array/slice as well.